builtin/logical/database/backend.go

```diff
@@ -286,3 +485,333 @@ cassandra, mssql, mysql, postgres
+			b.rotateCredentials(ctx, s)
+
+		case <-ctx.Done():
+			b.logger.Info("stopping periodic ticker")
+			return
+		}
+	}
+}
+
+// setCredentialsWAL is used to store information in a WAL that can retry a
+// credential setting or rotation in the event of partial failure.
+type setCredentialsWAL struct {
+	NewPassword string
+	OldPassword string
+	RoleName    string
+	Username    string
+
+	LastVaultRotation time.Time
+	Statements        dbplugin.Statements
+}
+
+// rotateCredentials sets a new password for a static account. This method is
+// invoked in the runTicker method, which is in it's own go-routine, and invoked
+// periodically (approximately every 5 seconds).
+//
+// This method loops through the priority queue, popping the highest priority
+// item until it encounters the first item that does not yet need rotation,
+// based on the current time.
+func (b *databaseBackend) rotateCredentials(ctx context.Context, s logical.Storage) error {
+	for {
+		item, err := b.credRotationQueue.Pop()
+		if err != nil {
+			if err == queue.ErrEmpty {
+				return nil
+			}
+			return err
+		}
+
+		// validate the role still exists
+		role, err := b.StaticRole(ctx, s, item.Key)
+		if err != nil {
+			b.logger.Warn("unable load role", "role", item.Key, "error", err)
+			continue
+		}
+		if role == nil {
+			b.logger.Warn("role not found", "role", item.Key, "error", err)
+			continue
+		}
+
+		// if "now" is greater than the Item priority, then it needs to be rotated
+		if time.Now().Unix() > item.Priority {
+			input := &setStaticAccountInput{
+				RoleName: item.Key,
+				Role:     role,
+			}
+
+			// If there is a WAL entry related to this Role, the corresponding WAL ID
+			// should be stored in the Item's Value field.
+			if walID, ok := item.Value.(string); ok {
+				walEntry := b.findStaticWAL(ctx, s, walID)
+				if walEntry != nil && walEntry.NewPassword != "" {
+					input.Password = walEntry.NewPassword
+					input.WALID = walID
+				}
+			}
+
+			resp, err := b.setStaticAccount(ctx, s, input)
+			if err != nil {
+				b.logger.Warn("unable to rotate credentials in periodic function", "error", err)
+				// Increment the priority enough so that the next call to this method
+				// likely will not attempt to rotate it, as a back-off of sorts
+				item.Priority = time.Now().Add(10 * time.Second).Unix()
+
+				// preserve the WALID if it was returned
+				if resp.WALID != "" {
+					item.Value = resp.WALID
+				}
+
+				if err := b.pushItem(item); err != nil {
+					b.logger.Warn("unable to push item on to queue", "error", err)
+				}
+				// go to next item
+				continue
+			}
+
+			lvr := resp.RotationTime
+			if lvr.IsZero() {
+				lvr = time.Now()
+			}
+
+			// update priority and push updated Item to the queue
+			nextRotation := lvr.Add(role.StaticAccount.RotationPeriod)
+			item.Priority = nextRotation.Unix()
+			if err := b.pushItem(item); err != nil {
+				b.logger.Warn("unable to push item on to queue", "error", err)
+			}
+		} else {
```

Comment:
Little nit pick, I think inverting this logic so that the base case is at the top will be a little cleaner

Reply:
Done